src/index.js

```diff
@@ -183,20 +183,56 @@ export default function (babel) {
 
 
   // HELPER FUNCTIONS
+  function visitTailExpressions(path, allowLoops, visitor) {
+    path.resync();
 
-  function transformTerminalExpressionsIntoArrPush(path, arrId) {
-    path.resync(); // not sure if this is necessary... c/p from addImplicitReturns
+    let tailPaths = getTailExpressions(path.get("body"), allowLoops);
+    for (let tailPath of tailPaths) {
+      if (tailPath.isExpressionStatement()) {
+        visitor(tailPath);
+      } else if (tailPath.isFunctionDeclaration()) {
+        // Promote function declarations to exprs.
```

Comment:
maybe I'm missing something... could help if you can show a test case that fails with the FunctionExpression transform removed...